pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go

```diff
@@ -235,33 +243,81 @@ func nodesWherePreemptionMightHelp(nodes []*framework.NodeInfo, m framework.Node
 	return potentialNodes
 }
 
+// calculateNumCandidates returns the number of candidates the findCandidates
+// method must produce from dry running based on the constraints given by
+// maxCandidatesAsPercentageOfNodes and minCandidates.
+func calculateNumCandidates(numNodes int) int {
+	if numNodes < minCandidates || maxCandidatesAsPercentageOfNodes >= 100 {
+		return numNodes
+	}
+	n := (numNodes * maxCandidatesAsPercentageOfNodes) / 100
+	if n < minCandidates {
+		return minCandidates
+	}
+	return n
+}
+
+// candidateOffset maintains state about the last window used for preemption,
+// calculates a new window each time, and updates its internal state.
+type candidateOffset struct {
+	offset int
+}
+
+// calculateOffset returns an <offset> and <numCandidates> that the caller must
+// use to find candidates. A variable offset is used as the starting index when
+// dry running preemption so that the same set of nodes don't get preempted
+// every time. If <offset> + <numCandidates> runs past the number of nodes, the
+// caller is expected to wrap around the node list.
+func (co *candidateOffset) calculateOffset(numNodes int) (int, int) {
+	if numNodes < minCandidates {
+		return 0, numNodes
+	}
+	offset := co.offset
+	numCandidates := calculateNumCandidates(numNodes)
+	if co.offset >= numNodes {
+		// This will never happen unless numNodes is lower than what we saw last
+		// time (cluster resize?) and the previous offset is past numNodes. Reset
+		// the counter and restart from the beginning.
```

Comment:
I guess this will happen normally since the potential nodes are based on each pod's filter status:

`https://github.com/kubernetes/kubernetes/pull/94814/commits/80c36642e2d55f9707b210a76bd7120b30b098e5#diff-7ee2d8038590ef2f75c2bfb4e5a4e944R174`

I'm also thinking of the `offset` should be meaningful for the same set of potential nodes, but actually, we will get different set of potential nodes for each pod.

Reply:
> Perhaps I misunderstood your comment, could you please elaborate what you meant by "offset should be meaningful for the same set of potential nodes"?

I mean, for example, we have 4 nodes: A, B ,C, D. then we try A and B first time, we have offset to record that next start from C. Then next around will try C and D. But due to the potential nodes are based on filter status. So maybe for PodA, we got A and B as potential  nodes only. After try the A and B, the offset back to 0. But for PodB, we got A, B, C and D as potential nodes.  Since offset is 0, then we try A and B again. Sorry for not describe clearly in the beginning :)

Reply:
I understand what you mean now. I agree that could happen, but I also think that scenario is less common than the case that round-robin solves. Round-robin guarantees that each node will have been tried for preemption in the average case after a certain number of pods.

(average case = nearly all nodes are available for preemption)